### PR TITLE
[SPARK-45680][CONNECT] Release session

### PR DESCRIPTION
**Pull request by juliuszsompolski**
### What changes were proposed in this pull request?

Introduce a new `ReleaseSession` Spark Connect RPC, which cancels everything running in the session and removes the session server side. Refactor code around managing the cache of sessions into `SparkConnectSessionManager`.

### Why are the changes needed?

Better session management.

### Does this PR introduce _any_ user-facing change?

Not really. `SparkSession.stop()` API already existed on the client side. It was closing the client's network connection, but the Session was still there cached for 1 hour on the server side.

Caveats, which were not really supported user behaviour:
* After `session.stop()`, user could have created a new session with the same session_id in Configuration. That session would be a new session on the client side, but connect to the old cached session in the server. It could therefore e.g. access that old session's state like views or artifacts.
* If a session timed out and was removed in the server, it used to be that a new request would re-create the session. The client would then see this as the old session, but the server would see a new one, and e.g. not have access to old session state that was removed.
* User is no longer allowed to create a new session with the same session_id as before.

### How was this patch tested?

Tests added.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by hvanhovell — COMMENTED** [review]
Looks good. Needs tests :)

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala
Comment (heyihong) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:670]: Should releaseSession be called inside `client.shutdown()`?

File: python/pyspark/sql/connect/session.py
Comment (heyihong) [inline python/pyspark/sql/connect/session.py:657]: ditto

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala
Comment (juliuszsompolski) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:670]: Since client is an internal API that shouldn't be used to end user, I deliberately left it to the session layer to call release session. Here this allows me to test some internal behaviours with the client in the tests. In the future, I imagine this could be useful in some situations where we may need to exchange a client / get a new GRPC channel without closing the session.

**Comment by hvanhovell** [comment]
@HyukjinKwon is this the same python test issue?

**Comment by HyukjinKwon** [comment]
I think we might need to retrigger ... https://github.com/juliuszsompolski/apache-spark/actions/runs/6695859696/job/18192423072

File: python/pyspark/sql/tests/connect/test_connect_basic.py
Comment (juliuszsompolski) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:3454]: @HyukjinKwon could you suggest how to help with it?
The issue is that this test uses a bunch of sessions with dummy hosts, but these sessions used to never issue any RPCs, but now they do ReleaseSession in the stop, and then it notices that the host doesn't work and it fails...
I cannot simply remove the stop, because these tests actually test how stop()
In scala I resolved it with a bit of hacks in https://github.com/apache/spark/pull/43546/files#diff-a2b06ed831053e086e47d191c1e86d26da81c2d2a5cbaedd34a712297a2b418dR126
https://github.com/apache/spark/pull/43546/files#diff-a2b06ed831053e086e47d191c1e86d26da81c2d2a5cbaedd34a712297a2b418dR126 ...

Comment (HyukjinKwon) [reply python/pyspark/sql/tests/connect/test_connect_basic.py:3454]: @juliuszsompolski I will take a look soon. Feel free to skip the tests for now, and we can just merge this first if this is an issue in tests only

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala
Comment (HyukjinKwon) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:670]: I wonder if we should make it stoppable/closable ... even for dummy hosts, etc.

Comment (HyukjinKwon) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:670]: Or, should we explicitly connect when we create a session? Let me match the fix to this for now in any event.

**Comment by HyukjinKwon** [comment]
I pushed some changes but I think you should take a look ...

**Comment by HyukjinKwon** [comment]
hmmm .. `test_error_enrichment_jvm_stacktrace` seems a real test failure too. lemme take a look soon

**Comment by HyukjinKwon** [comment]
The test should pass now.

**Comment by HyukjinKwon** [comment]
rebased

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
